clang/test/SemaSYCL/forward-decl.cpp

```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -x c++ -DNOVIRTUAL -fsycl-is-device -std=c++11 -fsyntax-only -verify -pedantic %s
+// RUN: %clang_cc1 -x c++ -DVIRTUAL -fsycl-is-device -std=c++11 -fsyntax-only -verify -pedantic %s
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+#ifdef NOVIRTUAL
+  kernel_single_task<class kernel_function_1>([]() {
+      // expected-no-diagnostics
+      class Foo *F;
+  });
+#elif VIRTUAL
+  kernel_single_task<class kernel_function_2>([]() {
+      // expected-error@+2{{No class with a vtable can be used in a SYCL kernel or any code included in the kernel}}
+      // expected-note@+1{{used here}}
+      class Boo {
+      public:
+        virtual int getBoo() { return 42; }
+      };
```

Comment:
On the other hands this class is not really used, so it might not worth an error.

Reply:
I agree with you, since spec says "no virtual member functions are allowed to be called in a SYCL kernel" and there are no function calls. But I believe that it's another issue, this patch is only about forward declaration resulting clang to crash (see test case under NOVIRTUAL guard).

Reply:
I think this test case is not about your patch.
You need to test that you won't generate invalid SYCL code if you used forward declaration and actual definition of your class contains something illegal for SYCL.
I guess our diagnostic system is not able to do it now but I'm trying to say that **IT'S NOT RIGHT** always say **THAT FORWARD DECLARED CLASS ALWAYS IS OK**.

Reply:
> You need to test that you won't generate invalid SYCL code if you used forward declaration and actual definition of your class contains something illegal for SYCL.

There is nothing wrong with forward _**declarations**_.
We have a few tests checking that compiler rejects invalid _**definitions**_.

Reply:
Okay. But allowing user use forward declarations still doesn't seem a good solution for me. IMO using only some pointer of forward declared class is not a real use case for forward declarations. I'm sure that user could try something like this:

a.cpp:
```
struct st {
  void do_something() { ... };
};

struct st* create_st() { ... };
void use_st(struct st* s) { s->do_something() };
```
b.cpp:
```
struct st* create_st();
void use_st(struct st* s);

void kernel() {
  struct st *s = create_st(); // First, we won't add definition of create_st to the device code.
  use_st(s);
}
```
Of course SYCL spec says

> The default behavior in SYCL applications is that all the definitions and declarations of the functions and methods are available to the SYCL compiler, in the same translation unit. When this is not the case, all the symbols that need to be exported to a SYCL library or from a C++ library to a SYCL application need to be defined using the macro: SYCL_EXTERNAL.

But we don't provide `SYCL_EXTERNAL` now. So with this patch we allow to user produce link fails but don't provide way to fix it. With our design of asynchronous exceptions handling user in addition won't see any linkage errors, just silent computation fail if he/she didn't use async_handler.

I'm ok with allowing forward declarations if we will support `SYCL_EXTERNAL` in the future.